app/models/user.rb

```diff
@@ -514,8 +514,14 @@ def active_practice
     active_practices.first.id
   end
 
-  def follow(other_user)
+  def follow(other_user, watch)
     following << other_user
+    change_watching(other_user, watch)
+  end
```

Comment:
【MUST】
これはトランザクションを張るべき処理だと思います
- どこまで説明したらいいか分からなくて冗長な説明になってます。よく分からないことがあれば聞いてください
- トランザクションは難しく、自分も自信がないです

---
複数の保存処理をまとめて行うとき、その処理の一部のみが実行され、残りが実行されない、という状況を避けたいことがあります
このような場合には保存処理を `transaction` で括り、完全に実行されるか、更新処理が起こる前の状態に戻る（ロールバックする）か、どちらかになるようにします

参考：https://qiita.com/tomokiyao/items/8697b523a0a5b9cf0bec

今回の場合、以下の二つの保存処理は途中で実行が止まると良くない種類のものに思えます

- ユーザーをフォローする
- ユーザーをwatchする、もしくはwatchを外す

ということで、この一連の処理をtransactionで囲いたいわけです
（ユーザーをwatchする、もしくはwatchを外す処理が失敗することはないだろうと踏んで、一々transactionで囲わないこともあるみたいですが、ここらへんの塩梅は全然わからない）

---
で、どうやってtransactionを書くんだという話ですが、以下のようにすればいいと思います

```suggestion
  def follow(other_user, watch)
    all_success = true
    ActiveRecord::Base.transaction do
      all_success &= following << other_user # ①
      all_success &= change_watching(other_user, watch) # ①
      unless all_success # ②
        raise ActiveRecord::Rollback
      end
    end
    all_success # ③
  end
```

参考: https://speakerdeck.com/jnchito/number-rubykansai-2018-01-13?slide=43

①どこかの保存処理に失敗すると、all_successはfalseになります。そうでなければtrueのままです
②保存処理に失敗していると、 `ActiveRecord::Rollback` という例外がraiseされ、ロールバックが発生します。transactionブロックによって例外はrescueされ、処理が続行します
③戻り値のall_successは保存が成功していればtrueに、失敗していればfalseになります

これでロールバックを可能にしつつ、 `true/false` を返すメソッドが書けていると思います

Reply:
どちらもfollowingモデルの更新だから、わざわざトランザクションを使わないでも、更新を1回だけ行うようにしたほうがいいのかなあ……

```ruby
following.create(followed: other_user, watch: watch)
```
（動くか確かめてないのであくまでイメージ）
でも `change_watching` メソッドが使い回せないのが嫌なのか……

とりあえずは上に書いたようなトランザクションで囲むスタイルで良いと思います

Reply:
すみません、方針代わって申し訳ないんですが、やっぱりこっちのほうが良いと思います

```suggestion
  def follow(other_user, watch)
    active_relationships.create(followed_id: other_user.id, watch: watch)
  end
```
参考: https://railsguides.jp/association_basics.html#has-manyで追加されるメソッド
`collection.create(attributes = {})` を使っています

transactionを書いちゃうと大仰ですし、テストも面倒になりますしね…（ロールバックされることを確認したほうがいいので）
`change_watching` は使えなくなってしまいますが、あまり大きな処理は書いてないですし、無理に使わなくてもいいかなと

Reply:
`active_relationships.create`を使えば1回の処理で済むんですね、なるほど……！
`User#follow`の書き方がずっと納得いっていなかったのでスッキリしました。

transactionについても今まで意識していなかったので勉強になります！
説明とても分かりやすいです〜ありがとうございます🙏